core/src/main/kotlin/com/strumenta/kolasu/ids/NodeIdProvider.kt

```diff
@@ -10,34 +9,44 @@ import com.strumenta.kolasu.model.Node as KNode
  * It is important that the logic is implemented so that given the same Node, the same ID is returned.
  */
 interface NodeIdProvider {
-    /**
-     * @param coordinates the coordinates at which the node is to be placed
-     * @return a valid LionWeb Node ID
-     */
-    fun idUsingCoordinates(kNode: KNode, coordinates: Coordinates): String
+    fun id(kNode: KNode): String
+    var parentProvider: NodeIdProvider?
+    val topLevelProvider: NodeIdProvider
+        get() = if (parentProvider == null) this else parentProvider!!.topLevelProvider
+}
 
-    fun id(kNode: KNode, overriddenCoordinates: Coordinates? = null): String {
-        return idUsingCoordinates(kNode, overriddenCoordinates ?: calculatedCoordinates(kNode))
-    }
+abstract class BaseNodeIdProvider : NodeIdProvider {
+    override var parentProvider: NodeIdProvider? = null
+        set(value) {
+            field = if (value == this) {
+                null
+            } else {
+                value
+            }
+        }
+}
+
+/**
+ * The common approach for calculating Node IDs is to use Semantic Node IDs where present,
+ * and Positional Node IDs in the other cases.
+ */
+class CommonNodeIdProvider(val semanticIDProvider: SemanticNodeIDProvider = DeclarativeNodeIdProvider()) :
+    BaseNodeIdProvider() {
 
-    private fun calculatedCoordinates(kNode: Node): Coordinates {
-        return if (kNode.parent == null) {
-            RootCoordinates
+    override fun id(kNode: Node): String {
+        return if (semanticIDProvider.hasSemanticIdentity(kNode)) {
+            semanticIDProvider.semanticID(kNode)
         } else {
-            NonRootCoordinates(this.id(kNode.parent!!), kNode.containingProperty()!!.name)
+            positionalID(kNode)
         }
     }
-}
 
-interface IDLogic {
-    fun calculatedID(coordinates: Coordinates): String
+    private fun positionalID(kNode: Node): String {
+        return StructuralNodeIdProvider().apply { parentProvider = this }.caching().id(kNode)
+    }
 }
 
-sealed class Coordinates
-
-object RootCoordinates : Coordinates()
-
-data class NonRootCoordinates(
-    val containerID: String,
-    val containmentName: String
-) : Coordinates()
+interface SemanticNodeIDProvider {
```

Comment:
What is a SemanticNodeId? What semantics are we referring to here?

Reply:
The idea is that some nodes have an identity that is relevant for the semantic of the language, which typically means "one can refer to them". For example, Class Declarations and Method Declarations have this "semantic identity". A package declaration or an expression have not.

The semantics API permit to attribute an ID to nodes that are relevant for symbol resolution, i.e., for nodes with a semantic identity. From the point of view of the symbol resolution other nodes do not need an id.

From the point of view of LionWeb every Node needs an ID, but for nodes which do not have a semantic identity the ID is based on where they are in the AST.

So we distinguish between nodes with a semantic ID (in practice, node that get an ID from the semantic API you created) and not which get a positional ID (so one calculated based on their position on the AST, so basically their coordinates)

Reply:
Thanks for the explanation.

In the symbol resolution API, we can define a _SymbolProvider_ to specify which nodes should be visible from the outside and what should be visible - i.e. a _SymbolDescription_. The identifier is stored in _SymbolDescription_ instances to have the possibility to retrieve the full _Node_ instance from the repository - i.e. _LionWebRepository_ for example. 

In this context, I think we could only keep the positional identifier. Nodes that we do not want to make visible from the outside - e.g. expressions - will not have a corresponding _SymbolDescription_ and will not be stored in the _SymbolRepository_. What do you think?

When creating a symbol description, we can then use one of these IdProvider classes - perhaps an interface if there is any or directly the default identifier provider that we plan to use. The only feature that is expected by the symbol provider and other semantics components is to be able to generally obtain the same ID for a given node in a deterministic way. I think this is covered by your implementation. 

To summarise, I think we could avoid having an additional concept of semantic identifier. Am I missing something?

Reply:
> Nodes that we do not want to make visible from the outside - e.g. expressions - will not have a corresponding SymbolDescription and will not be stored in the SymbolRepository. What do you think?

Yes, absolutely.

But before you write "In this context, I think we could only keep the positional identifier", while it would be the nodes with semantic IDs and not positional IDs to end up in the Symbol Repository

> When creating a symbol description, we can then use one of these IdProvider classes - perhaps an interface if there is any or directly the default identifier provider that we plan to use. 

Actually I was thinking the other way around: you have a system to calculate the semantic IDs. The more general NodeIdProvider should first call your system and if your system can produce an ID that would be used. Otherwise, the positional ID would be generated. This is what the class `CommonNodeIdProvider` does, trying to call `DeclarativeNodeIdProvider` first. When `DeclarativeNodeIdProvider` does not return an ID, because we passed to it an expression, then the positional ID would be calculated.